incubator/orm/natural_key.go

```diff
@@ -3,101 +3,97 @@ package orm
 import (
 	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
 )
 
 var _ TableBuilder = &naturalKeyTableBuilder{}
 
-type naturalKeyer func(val interface{}) []byte // todo: note: in the api design this does not return an error unlike other indexer functions do
-
-func NewNaturalKeyTableBuilder(prefixData, prefixSeq, prefixIndex byte, key sdk.StoreKey, cdc *codec.Codec, model interface{}, getPrimaryKey naturalKeyer) *naturalKeyTableBuilder {
+func NewNaturalKeyTableBuilder(prefixData, prefixSeq, prefixIndex byte, key sdk.StoreKey, cdc *codec.Codec, model NaturalKeyed) *naturalKeyTableBuilder {
 	if prefixIndex == prefixData || prefixData == prefixSeq {
 		panic("prefixIndex must be unique")
 	}
 
 	builder := NewAutoUInt64TableBuilder(prefixData, prefixSeq, key, cdc, model)
 
 	idx := NewUniqueIndex(builder, prefixIndex, func(value interface{}) (bytes [][]byte, err error) {
-		return [][]byte{getPrimaryKey(value)}, nil
+		obj, ok := value.(NaturalKeyed)
+		if !ok {
+			return nil, errors.Wrapf(ErrType, "%T", value)
+		}
+		return [][]byte{obj.NaturalKey()}, nil
```

Comment:
Looking at this, I have two questions for API design:

1. Should we pass errors, of just go sdk-style with panics, which turns this into a one liner: `return [][]byte{value.(NaturalKeyed).NaturalKey()}`

2. Unique indexes don't accept multi-values, right? So couldn't we change the function signature of `NewUniqueIndex` to take `func(value inteface{}) ([]byte, error)` - a single byte slice, rather than slice of slices? (Of course I may misunderstand the functionality, but this is my first impression reading the usage)

Reply:
About 1. : I have mixed feelings. I dropped error return values already where they did not make any sense. But I did not optimize for "panic"  😕 
About 2: good point!  Will do